core/src/main/java/org/apache/spark/shuffle/api/ShufflePartitionWriter.java

```diff
@@ -31,12 +30,10 @@
   OutputStream openPartitionStream();
 
   /**
-   * Indicate that the partition was written successfully and there are no more incoming bytes. Returns
-   * the length of the partition that is written. Note that returning the length is mainly for backwards
-   * compatibility and should be removed in a more polished variant. After this method is called, the writer
-   * will be discarded; it's expected that the implementation will close any underlying resources.
+   * Indicate that the partition was written successfully and there are no more incoming bytes.
+   * Returns a {@link CommittedPartition} indicating information about that written partition.
    */
-  long commitAndGetTotalLength();
+  CommittedPartition commitPartition();
```

Comment:
I merged my branch with your recent changes and I'm not a big fan of this.

I think it useful to have the API expose something like this, but your implementation has two very important issues:

- it's opinionated about what `CommittedPartition` is
- it doesn't allow the implementation to return null

There may be a third issue, which is what happens with very large shuffles, where you'll have a ton of these objects.

Reply:
What would you propose as an alternative?

Reply:
(Regarding the third issue, we have plans on how to compress these on the driver, such that we only store the set of unique locations and create an index to look up the location per reduce id for a given map output. But perhaps more efficient is making `ShuffleLocation` be an opaque `Array[Byte]`.)

Reply:
My original idea was for the map side to just return an opaque `Object` (or `Serializable`, I guess), and the reduce side to take that as an argument.

It might need some extra APIs, though (the driver must be able to re-organize things so that the reducers have a view of the data corresponding to all of the different mappers).

Something like:

```
ShuffleMapOutputWriter {
  Serializable commitAllPartitions() throws IOException
}

ShuffleDataIO {
  Serializable mergeMapOutput(Serializable mapOutput);
}

ShuffleReadSupport {
  ShufflePartitionReader newPartitionReader(String appId, int shuffleId, int mapId, Serializable metadata)
}
```

That's just a quick dump off the top of my head, so may not really be workable.

Reply:
(But, really, for my current use, just being able to return `null` and have things work would have been great.)